Check bound before reading past socket buffer.

shane-lawrence opened the pull request:
Signed-off-by: Shane Lawrence <shane@lawrence.dev>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines in the [CONTRIBUTING.md](https://github.com/falcosecurity/.github/blob/master/CONTRIBUTING.md) file and learn how to compile Falco from source [here](https://falco.org/docs/source).
2. Please label this pull request according to what type of issue you are addressing.
3. Please add a release note!
4. If the PR is unfinished while opening it specify a wip in the title before the actual title, for example, "wip: my awesome feature"
-->

**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

> /kind feature

<!--
Please remove the leading whitespace before the `/kind <>` you uncommented.
-->

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area API-version

> /area build

> /area CI

> /area driver-kmod

> /area driver-bpf

> /area driver-modern-bpf

> /area libscap-engine-bpf

> /area libscap-engine-gvisor

> /area libscap-engine-kmod

> /area libscap-engine-modern-bpf

> /area libscap-engine-nodriver

> /area libscap-engine-noop

> /area libscap-engine-source-plugin

> /area libscap-engine-savefile

> /area libscap

> /area libpman

> /area libsinsp

> /area tests

> /area proposals

<!--
Please remove the leading whitespace before the `/area <>` you uncommented.
-->

**Does this PR require a change in the driver versions?**

> /version driver-API-version-major

> /version driver-API-version-minor

> /version driver-API-version-patch

> /version driver-SCHEMA-version-major

> /version driver-SCHEMA-version-minor

> /version driver-SCHEMA-version-patch

<!--
Please remove the leading whitespace before the `/version <>` you uncommented.
-->

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
If PR is `kind/failing-tests` please post the related issues/tests in a comment and do not use `Fixes`.
-->

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

<!--
If no, you have to do nothing.
If yes, a release note is required:
Delete `NONE` and enter your extended release note in the block below.
Please note, the release note follows the "conventional commit specification" (https://www.conventionalcommits.org/en/v1.0.0/):
For example: `fix: broken link`.
If the PR requires additional action from users switching to the new release, prepend the string "action required:".
For example, `action required: change the API interface of libscap`.
-->

```release-note
NONE
```

poiana commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/2271#" title="Author self-approved">shane-lawrence</a>*
**Once this PR has been reviewed and has the lgtm label**, please assign [molter73](https://github.com/molter73) for approval. For more information see [the Kubernetes Code Review Process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process).

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["molter73"]} -->

poiana commented:
Welcome @shane-lawrence! It looks like this is your first PR to falcosecurity/libs 🎉

FedeDP commented:
Thanks for this contribution; it makes sense to me.
/milestone 0.21.0

poiana commented:
Thanks for your pull request. Before we can look at it, you'll need to add a 'DCO signoff' to your commits.

:memo: **Please follow instructions in the [contributing guide](https://github.com/falcosecurity/.github/blob/main/CONTRIBUTING.md) to update your commits with the DCO**

Full details of the Developer Certificate of Origin can be found at [developercertificate.org](https://developercertificate.org/).

**The list of commits missing DCO signoff**:

- [dd4443c](https://github.com/falcosecurity/libs/commits/dd4443c306bb09d7908b31026583ac39c709a66f) Test for overflow reading proc IPv4 buffer.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

shane-lawrence commented:
I added tests and confirmed that it triggers a segfault without the fix but succeeds with the fix.

shane-lawrence commented:
Just rebased on master to pick up the API changes @ekoops made recently.

leogr commented:
Hey @shane-lawrence 

Thank you for this PR! Just noticed :point_down: 
![image](https://github.com/user-attachments/assets/d58e1a1d-1e95-4cec-8633-809e85a50f50)
See https://github.com/falcosecurity/libs/actions/runs/13171693500/job/36769287438?pr=2271

May you fix the [code formatting](https://github.com/falcosecurity/libs/blob/master/Contributing.md#enforce-coding-style-), please?

LucaGuerra reviewed (dismissed):
Great Catch! Thank you!

poiana commented:
LGTM label has been added.  <details>Git tree hash: 431f76b8d8276d05cd2821e4cd7cc62d0faf0a32</details>

LucaGuerra commented:
I'm restarting the CI, if it passes it's good for me

poiana commented:
New changes are detected. LGTM label has been removed.

poiana commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/2271#" title="Author self-approved">shane-lawrence</a>*
**Once this PR has been reviewed and has the lgtm label**, please ask for approval from [lucaguerra](https://github.com/lucaguerra). For more information see [the Kubernetes Code Review Process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process).

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["lucaguerra"]} -->

poiana commented:
Thanks for your pull request. Before we can look at it, you'll need to add a 'DCO signoff' to your commits.

:memo: **Please follow instructions in the [contributing guide](https://github.com/falcosecurity/.github/blob/main/CONTRIBUTING.md) to update your commits with the DCO**

Full details of the Developer Certificate of Origin can be found at [developercertificate.org](https://developercertificate.org/).

**The list of commits missing DCO signoff**:

- [46af469](https://github.com/falcosecurity/libs/commits/46af469a5874651d89533ce4fbbae3906ddbdb2c) Add test for buffer overrun.
- [9c09739](https://github.com/falcosecurity/libs/commits/9c09739ce9ac6841d8d9163284be8021878e7d81) Reintroduce bug to test the test.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

leogr commented:
CI restarted!

shane-lawrence commented:
Looks like libscap_test [isn't running](https://github.com/falcosecurity/libs/actions/runs/13246652266/job/36975317472) and [wouldn't detect memory issues](https://github.com/falcosecurity/libs/blob/5e06e37aaacd1c45b50bc17af3d6af51098aa58a/.github/workflows/drivers_ci.yml#L76C1-L78C1) even if it was. I opened [a separate issue](https://github.com/falcosecurity/libs/issues/2276) for that to get the regression test working.

Here is the result of running the test locally, with AddressSanitizer, without the scap_fds.c fix:
```
Running main() from /usr/src/falcosecurity/libs/build/googletest-src/googletest/src/gtest_main.cc
Note: Google Test filter = scap_fds.buffer_overflow_test
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from scap_fds
[ RUN      ] scap_fds.buffer_overflow_test
=================================================================
==485201==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x72f9b87ff800 at pc 0x5f7aa9dea91d bp 0x7ffd073cbf30 sp 0x7ffd073cbf20
READ of size 1 at 0x72f9b87ff800 thread T0
    #0 0x5f7aa9dea91c in scap_fd_read_ipv4_sockets_from_proc_fs /usr/src/falcosecurity/libs/userspace/libscap/linux/scap_fds.c:769
    #1 0x5f7aa9defe51 in test_time_wait_socket_at_buffer_end /usr/src/falcosecurity/libs/test/libscap/test_suites/userspace/linux/scap_fds_impl.c:38
    #2 0x5f7aa9de3749 in scap_fds_buffer_overflow_test_Test::TestBody() /usr/src/falcosecurity/libs/test/libscap/test_suites/userspace/linux/scap_fds.cpp:9
    #3 0x5f7aa9ef0810 in void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /usr/src/falcosecurity/libs/build/googletest-src/googletest/src/gtest.cc:2599
    #4 0x5f7aa9ef0810 in void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /usr/src/falcosecurity/libs/build/googletest-src/googletest/src/gtest.cc:2635
    #5 0x5f7aa9ed7c2a in testing::Test::Run() /usr/src/falcosecurity/libs/build/googletest-src/googletest/src/gtest.cc:2674
    #6 0x5f7aa9ed8009 in testing::Test::Run() /usr/src/falcosecurity/libs/build/googletest-src/googletest/src/gtest.cc:2665
    #7 0x5f7aa9ed8009 in testing::TestInfo::Run() /usr/src/falcosecurity/libs/build/googletest-src/googletest/src/gtest.cc:2853
    #8 0x5f7aa9ed81d0 in testing::TestSuite::Run() /usr/src/falcosecurity/libs/build/googletest-src/googletest/src/gtest.cc:3012
    #9 0x5f7aa9ee60fe in testing::TestSuite::Run() /usr/src/falcosecurity/libs/build/googletest-src/googletest/src/gtest.cc:2986
    #10 0x5f7aa9ee60fe in testing::internal::UnitTestImpl::RunAllTests() /usr/src/falcosecurity/libs/build/googletest-src/googletest/src/gtest.cc:5870
    #11 0x5f7aa9ed866d in bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /usr/src/falcosecurity/libs/build/googletest-src/googletest/src/gtest.cc:2599
    #12 0x5f7aa9ed866d in bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /usr/src/falcosecurity/libs/build/googletest-src/googletest/src/gtest.cc:2635
    #13 0x5f7aa9ed866d in testing::UnitTest::Run() /usr/src/falcosecurity/libs/build/googletest-src/googletest/src/gtest.cc:5444
    #14 0x5f7aa9db2793 in RUN_ALL_TESTS() /usr/src/falcosecurity/libs/build/googletest-src/googletest/include/gtest/gtest.h:2293
    #15 0x5f7aa9db2793 in main /usr/src/falcosecurity/libs/build/googletest-src/googletest/src/gtest_main.cc:51
    #16 0x72f9bb82a1c9 in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58
    #17 0x72f9bb82a28a in __libc_start_main_impl ../csu/libc-start.c:360
    #18 0x5f7aa9dbbef4 in _start (/usr/src/falcosecurity/libs/build/test/libscap/libscap_test+0xd8ef4) (BuildId: 75a8a638aa36547e71442810f6545d067b31c57a)

0x72f9b87ff800 is located 0 bytes after 1048576-byte region [0x72f9b86ff800,0x72f9b87ff800)
allocated by thread T0 here:
    #0 0x72f9bc0fd9c7 in malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:69
    #1 0x5f7aa9de9507 in scap_fd_read_ipv4_sockets_from_proc_fs /usr/src/falcosecurity/libs/userspace/libscap/linux/scap_fds.c:676

SUMMARY: AddressSanitizer: heap-buffer-overflow /usr/src/falcosecurity/libs/userspace/libscap/linux/scap_fds.c:769 in scap_fd_read_ipv4_sockets_from_proc_fs
Shadow bytes around the buggy address:
  0x72f9b87ff580: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x72f9b87ff600: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x72f9b87ff680: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x72f9b87ff700: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x72f9b87ff780: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x72f9b87ff800:[fa]fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x72f9b87ff880: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x72f9b87ff900: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x72f9b87ff980: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x72f9b87ffa00: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x72f9b87ffa80: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==485201==ABORTING
```

And here it is when I apply the fix in this PR:
```
[ 94%] Linking CXX executable libscap_test
[100%] Built target libscap_test
Running main() from /usr/src/falcosecurity/libs/build/googletest-src/googletest/src/gtest_main.cc
Note: Google Test filter = scap_fds.buffer_overflow_test
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from scap_fds
[ RUN      ] scap_fds.buffer_overflow_test
[       OK ] scap_fds.buffer_overflow_test (12 ms)
[----------] 1 test from scap_fds (12 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (13 ms total)
[  PASSED  ] 1 test.

=================================================================
==487428==ERROR: LeakSanitizer: detected memory leaks

Indirect leak of 8768480 byte(s) in 7829 object(s) allocated from:
    #0 0x7db71ccfd9c7 in malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:69
    #1 0x5b3b0cf415b9 in scap_fd_read_ipv4_sockets_from_proc_fs /usr/src/falcosecurity/libs/userspace/libscap/linux/scap_fds.c:699

Indirect leak of 65536 byte(s) in 1 object(s) allocated from:
    #0 0x7db71ccfd9c7 in malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:69
    #1 0x5b3b0cf42315 in scap_fd_read_ipv4_sockets_from_proc_fs /usr/src/falcosecurity/libs/userspace/libscap/linux/scap_fds.c:809

Indirect leak of 64 byte(s) in 1 object(s) allocated from:
    #0 0x7db71ccfd9c7 in malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:69
    #1 0x5b3b0cf42175 in scap_fd_read_ipv4_sockets_from_proc_fs /usr/src/falcosecurity/libs/userspace/libscap/linux/scap_fds.c:809

SUMMARY: AddressSanitizer: 8834080 byte(s) leaked in 7831 allocation(s).
```

You can see that it resolves the buffer overrun and the test passes. It also flags the other issue (with ino=0) that I mentioned in Slack. That's out of scope for this PR but I hope to open a separate one tonight to resolve it.

Everything here is ready to go as far as I can tell.

FedeDP commented:
> Looks like libscap_test [isn't running](https://github.com/falcosecurity/libs/actions/runs/13246652266/job/36975317472) a

Indeed it is: https://github.com/falcosecurity/libs/blob/5e06e37aaacd1c45b50bc17af3d6af51098aa58a/.github/workflows/drivers_ci.yml#L103, or am i missing anything?

> Here is the result of running the test locally, with AddressSanitizer, without the scap_fds.c fix:

Nice!

> Everything here is ready to go as far as I can tell.

Thanks! And of course thanks for taking the time to add a test! I agree with you that we should definitely run the libscap_test with asan enabled.

FedeDP commented on test/libscap/CMakeLists.txt:
```diff
@@ -44,9 +44,24 @@ set(LIBSCAP_TESTS_LIBRARIES "${GTEST_LIB}" "${GTEST_MAIN_LIB}" "${CMAKE_THREAD_L
 set(LIBSCAP_TESTS_SOURCES ${USERSPACE_TEST_SUITE})
 
+# Add _GNU_SOURCE definition for makedev and set C flags
+add_definitions(-D_GNU_SOURCE)
```
This would add the definition globally, right? What about using https://cmake.org/cmake/help/latest/command/target_compile_definitions.html ?

FedeDP replied:
Can we remove this line then?

FedeDP commented on test/libscap/CMakeLists.txt:
```diff
@@ -44,9 +44,24 @@ set(LIBSCAP_TESTS_LIBRARIES "${GTEST_LIB}" "${GTEST_MAIN_LIB}" "${CMAKE_THREAD_L
+
+# Add scap_engine_util for Linux tests
+if(CMAKE_SYSTEM_NAME MATCHES "Linux")
+	list(APPEND LIBSCAP_TESTS_LIBRARIES scap_engine_util)
```
Why was this line moved above?

shane-lawrence replied:
Fixed in https://github.com/falcosecurity/libs/pull/2271/commits/3effc75f9fef475771f6eb25b4cfa716fa49caf3.

FedeDP commented on test/libsinsp_e2e/vtidcollision.c:
```diff
@@ -16,7 +16,10 @@ limitations under the License.
 
 */
 
+#ifndef _GNU_SOURCE
```
If we don't use `add_definitions(-D_GNU_SOURCE)` above, we can avoid this small patch.

shane-lawrence replied:
Fixed in https://github.com/falcosecurity/libs/pull/2271/commits/4cde899f228bed4b6e68e4ad7df31638047a11ed.

FedeDP commented on test/libscap/linux/CMakeLists.txt:
```diff
@@ -0,0 +1,5 @@
+
+target_link_libraries(unit-test-libscap-linux "${GTEST_LIB}" "${GTEST_MAIN_LIB}" scap)
+
+add_test(NAME unit-test-libscap-linux COMMAND unit-test-libscap-linux)
```
Who is going to run `unit-test-libscap-linux` ? Can we somehow link it to `libscap_test` to make it automatically run (but only on linux of course)?

FedeDP replied:
Oh they are already part of the libscap_test sources, sorry. So, this is just a small helper to only run linux tests?

shane-lawrence replied:
I added this when I first gave up on implementing the test fully in C++, before I decided on scap_fds_impl.c. It's not necessary and I've removed it in https://github.com/falcosecurity/libs/pull/2271/commits/2c250970606a0732728aed755db1ad1d89f7c6e4.

LucaGuerra commented:
The job did run AFAICT: https://github.com/falcosecurity/libs/actions/runs/13250940684/job/37009150158?pr=2271 but @shane-lawrence you're right noticing that ASan is not enabled there 🤦 . As @FedeDP mentioned, if you could add that switch it'd be appreciated 🙏

shane-lawrence commented:
> Please note that since drivers CI is pretty heavy, it only triggers when either userspace/libscap, userspace/libpman or driver folders are touched.

Oh that's it! When I ran the new test without the changes to libscap, [CI skipped libscap_test](https://github.com/falcosecurity/libs/actions/runs/13246652266/job/36975317472). I understand now that's because it didn't see any changes to userspace/libscap.

> I agree with you that we should definitely run the libscap_test with asan enabled; feel free to just add `-DUSE_ASAN=On -DUSE_UBSAN=On` to cmake configure options to the `test-scap` CI job :)

Great! I added this in https://github.com/falcosecurity/libs/pull/2271/commits/824363c1de77b44915a9cb5a76066477748ba9f3.

FedeDP commented:
Oh nice, enabling asan found another issue in the test-scap :laughing: 
```
[ RUN      ] scap_event.empty_clone
/home/runner/work/libs/libs/test/libscap/test_suites/userspace/scap_event.cpp:107:2: runtime error: reference binding to misaligned address 0x6030000056c6 for type 'const unsigned int', which requires 4 byte alignment
0x6030000056c6: note: pointer points here
 00 00 de 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  00 00
```

FedeDP commented on test/libscap/test_suites/userspace/linux/scap_test_sockets.txt:
Was this committed by error?

shane-lawrence replied:
No, it's mock data to allow testing. I couldn't use real data from a production host but we need input that meets the specific conditions for triggering the bug, so I generated this. It should mimic a valid sockets table and must have a line with consecutive spaces at the 1MB buffer boundary and the inode number just beyond it.

The mock data is read for the test in [scap_fds_impl.c:36](https://github.com/falcosecurity/libs/pull/2271/files/2c250970606a0732728aed755db1ad1d89f7c6e4#diff-d36e55c9f906b9a2bc1ce2fc7a52c059d370bdb0658f6a439b5982480339e2baR36).

shane-lawrence commented:
> About the test-scap tests failing, cc @LucaGuerra any idea?

Most of the tests that trigger this can avoid it by copying the field before evaluating them like this:
```
-       EXPECT_EQ(evt->nparams, 0);
+       uint16_t nparams_copy;
+       memcpy(&nparams_copy, &evt->nparams, sizeof(uint16_t));
+       EXPECT_EQ(nparams_copy, 0);
```

That won't work for test_scap_create_event though, since it's asserting that each field is in the correct location in the packed struct. I'm a bit out of my depth here, but it looks like allowing 2 bytes of padding would satisfy the alignment check on 4-byte boundaries, but it would also be a sweeping change and even a couple bytes can be significant. 

Would it be okay if I just suppress these for now and open a separate issue?

FedeDP commented:
> I'm a bit out of my depth here, but it looks like allowing 2 bytes of padding would satisfy the alignment check on 4-byte boundaries, but it would also be a sweeping change and even a couple bytes can be significant when it's on every event.

Yep we cannot change the struct padding :/ 

I am ok with your solution, but at that point it would be better to expose helper functions in scap_event.h like `get_ts(scap_event *evt, uint64_t *ts)`, since that is the correct way to access these fields (perhaps with a big comment :D )

> Would it be okay if I just suppress these for now and open a separate issue?

If you agree, i'd just suppress `ahhah` for now and fix the others.

Or, you can revert the enablement of ASAN in the libscap test CI and we will re-enable it in a follow up PR trying also to fix these failures :) That works too!

poiana commented:
Thanks for your pull request. Before we can look at it, you'll need to add a 'DCO signoff' to your commits.

:memo: **Please follow instructions in the [contributing guide](https://github.com/falcosecurity/.github/blob/main/CONTRIBUTING.md) to update your commits with the DCO**

Full details of the Developer Certificate of Origin can be found at [developercertificate.org](https://developercertificate.org/).

**The list of commits missing DCO signoff**:

- [0092fe5](https://github.com/falcosecurity/libs/commits/0092fe5474eea04bf7966e37346161493f65726a) Add getter for nparams.
- [007c4f2](https://github.com/falcosecurity/libs/commits/007c4f21a72dff6e658b4649ee9141eb845f0eb9) Use memcmp on raw bytes to avoid alignment issues.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

shane-lawrence commented:
I added a getter to safely check nparams, and changed the test to use memcmp with a uint8_t* so we can safely compare bytes in the event without memory alignment issues.

shane-lawrence commented:
Next I ran into:
```
[ RUN      ] scap_ppm_sc.scap_ppm_sc_from_name
/usr/src/falcosecurity/libs/build/googletest-src/googletest/include/gtest/gtest.h:1358:11: runtime error: load of value 4294967295, which is not a valid value for type 'ppm_sc_code'
```

I fixed this in 6e103ba9dc849e4dba384aef5bae901dc0d77e23.

FedeDP commented:
Since the test below (`scap_native_id_to_ppm_sc`) checks that:
```
ASSERT_EQ(scap_native_id_to_ppm_sc(80000000), PPM_SC_UNKNOWN);
ASSERT_EQ(scap_native_id_to_ppm_sc(-12), PPM_SC_UNKNOWN);
```
ie: a wrong syscall ID lead to `PPM_SC_UNKNOWN`, i think that it's ok that a wrong syscall name returns `PPM_SC_UNKNOWN` (and not -1).
I'd fix the `scap_ppm_sc_from_name` test to check for `PPM_SC_UNKNOWN` instead of `-1`.

Again, thanks you are doing a terrific job in this PR (and sorry if it is growing a little bit over your expectations!)

shane-lawrence commented:
Thanks for all of your help so far! I would like to get the undefined behaviours fixed but they're out of scope for this PR. All of the address sanitizer issues detected by the tests have been resolved and it will help to ensure that future changes to libscap don't reintroduce a similar tricky segfault. I'm leaving asan enabled but disabling ubsan and moving the alignment and enum fixes to a separate branch for a future PR.

While we can see that some potential bugs are ignored here, I I think this leaves us with a resolution to my issue and detection in place for a whole class of bugs, while keeping the PR to a reasonable scope. Let me know if you want me to rebase to clean up the commit history a bit, but hopefully the overall changes are now ready for a final review :).

FedeDP commented:
I fully agree, and am really sorry that what seemed to be a simple patch became much larger :/ 
Thanks for your hard work!

FedeDP reviewed (commented):
LGTM, left a small comment.
Also, do you mind squashing the changes in like 2 commits (the fix + the new test)?
Thanks, then we are good to go for real ahah

shane-lawrence commented:
I rebased on master with two clean commits and CI is running now. Let me know if you have any other suggestions.

FedeDP commented:
Everything :green_circle: on my side. Thank you very much once again!

poiana commented:
LGTM label has been added.  <details>Git tree hash: 5f821cb6085f2838efb01fc7576c398fa9df2fb7</details>
